mgr/volumes: Fix pool removal on volume deletion

kotreshhr opened the pull request:
While volume deletion, the associated pools are not always
removed. The pools are removed only if the volume is created
using mgr plugin and not if created with custom osd pools.
This is because mgr plugin generates pool names with specific
pattern. Both create and delete volume relies on it. This
patch fixes the issue by identifying the pools of the volume
without relying on the pattern.

Fixes: https://tracker.ceph.com/issues/45910
Signed-off-by: Kotresh HR <khiremat@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

joscollin commented on src/pybind/mgr/volumes/fs/operations/volume.py:
```diff
@@ -47,7 +65,7 @@ def create_volume(mgr, volname, placement):
+def delete_volume(mgr, volname, metadata_pool, data_pools):
     """
     delete the given module (tear down mds, remove filesystem)
     """
```
nit: update the comment mentioning delete pools.

kotreshhr replied:
Done

kotreshhr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -292,13 +295,18 @@ def test_volume_rm(self):
+                #check if osd pools are gone
+                osds = json.loads(self._raw_cmd("osd", "pool", "ls", "--format=json-pretty"))
+                for pool in vol_status["pools"]:
+                    self.assertNotIn(pool["name"], osds)
```
It was encouraged to use asserts rather than RuntimeError in previous PRs. Hence used assert.

kotreshhr commented:
> You're not really testing the behavior you're trying to fix: a volume with a data pool that is not generated (i.e. added out-of-band) is not deleted during volume removal.

The volumes created by test framework chooses the data pool names differently than the mgr plugin. Hence the data pool was never removed with the tests. So I didn't specifically create the osd and added out-of-band. 

@batrick 
Do you want me to specifically add the data pool as part of the test case ?

ajarr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -292,13 +295,18 @@ def test_volume_rm(self):
                     raise RuntimeError("Expected the 'fs volume rm' command to succeed. "
                                        "The volume {0} not removed.".format(self.volname))
+                #check if osd pools are gone
+                osds = json.loads(self._raw_cmd("osd", "pool", "ls", "--format=json-pretty"))
```
This bug is about not being able to remove volume's pools created by a non ceph-mgr interface. Where is this tested? 

You can maybe add a data pool to the mgr-created volume, then remove volume, and check whether all the volume's pools are removed.

ajarr replied:
OK. Looks like Patrick already pointed this issue out. You can use my above suggestion to fix this.

kotreshhr replied:
> > > You're not really testing the behavior you're trying to fix: a volume with a data pool that is not generated (i.e. added out-of-band) is not deleted during volume removal.
> > 
> > 
> > The volumes created by test framework chooses the data pool names differently than the mgr plugin.
> 
> I don't see where this happens. Can you point that out? In setUp() I only see self._create_or_reuse_test_volume() that doesn't use specially created pools.

I observed it on local teuthology run and assumed the same for actual test. Locally, it will destroy any existing fs and creates one as below.
```
$python ../qa/tasks/vstart_runner.py --kclient --interactive tasks.cephfs.test_volumes.TestVolumes.test_volume_rm
...
...
2020-06-22 22:39:02,653.653 INFO:tasks.cephfs.filesystem:Creating filesystem 'cephfs'
2020-06-22 22:39:02,654.654 INFO:__main__:Running ['./bin/ceph', 'osd', 'pool', 'create', 'cephfs_metadata', '8']
2020-06-22 22:39:03,420.420 INFO:__main__:Running ['./bin/ceph', 'osd', 'pool', 'create', 'cephfs_data', '8']

```
> 
> > Hence the data pool was never removed with the tests. So I didn't specifically create the osd and added out-of-band.
> 
> Why do you've to create OSDs? I guess you meant pools?
> 
> > @batrick
> > Do you want me to specifically add the data pool as part of the test case ?



> > > You're not really testing the behavior you're trying to fix: a volume with a data pool that is not generated (i.e. added out-of-band) is not deleted during volume removal.
> > 
> > 
> > The volumes created by test framework chooses the data pool names differently than the mgr plugin.
> 
> I don't see where this happens. Can you point that out? In setUp() I only see self._create_or_reuse_test_volume() that doesn't use specially created pools.
> 
> > Hence the data pool was never removed with the tests. So I didn't specifically create the osd and added out-of-band.
> 
> Why do you've to create OSDs? I guess you meant pools?
> 
> > @batrick
> > Do you want me to specifically add the data pool as part of the test case ?

ajarr commented on src/pybind/mgr/volumes/fs/operations/volume.py:
```diff
@@ -74,11 +92,16 @@ def delete_volume(mgr, volname):
+        r, outb, outs = remove_pool(mgr, data_pool)
+        if r != 0:
+            return r, outb, outs
+    result_str = "metadata pool: {0} data pool: {1} removed".format(metadata_pool, repr(data_pools))
```
Why repr() over str()?

kotreshhr replied:
No strong reason. I am used to use repr as it's more unambiguous. str works here. Fixed it.

ajarr commented:
> > You're not really testing the behavior you're trying to fix: a volume with a data pool that is not generated (i.e. added out-of-band) is not deleted during volume removal.
> 
> The volumes created by test framework chooses the data pool names differently than the mgr plugin.

I don't see where this happens. Can you point that out? In setUp() I only see self._create_or_reuse_test_volume() that doesn't use specially created pools. 

> Hence the data pool was never removed with the tests. So I didn't specifically create the osd and added out-of-band.
> 
Why do you've to create OSDs? I guess you meant pools?

> @batrick
> Do you want me to specifically add the data pool as part of the test case ?

kotreshhr commented:
> > > You're not really testing the behavior you're trying to fix: a volume with a data pool that is not generated (i.e. added out-of-band) is not deleted during volume removal.
> > 
> > 
> > The volumes created by test framework chooses the data pool names differently than the mgr plugin.
> 
> I don't see where this happens. Can you point that out? In setUp() I only see self._create_or_reuse_test_volume() that doesn't use specially created pools.
> 
I observed that on local teuthology run on vstart cluster and assumed the same could be the case with actual test. Locally it will destroy any existing fs and re-create one as below.
```
$python ../qa/tasks/vstart_runner.py --kclient --interactive tasks.cephfs.test_volumes.TestVolumes.test_volume_rm
...
...
2020-06-22 22:39:02,653.653 INFO:tasks.cephfs.filesystem:Creating filesystem 'cephfs'
2020-06-22 22:39:02,654.654 INFO:__main__:Running ['./bin/ceph', 'osd', 'pool', 'create', 'cephfs_metadata', '8']
2020-06-22 22:39:03,420.420 INFO:__main__:Running ['./bin/ceph', 'osd', 'pool', 'create', 'cephfs_data', '8']
2020-06-22 22:39:04,480.480 INFO:__main__:Running ['./bin/ceph', 'fs', 'new', 'cephfs', 'cephfs_metadata', 'cephfs_data', '--force']

```
I will change the test case and rebase the patch. 

> > Hence the data pool was never removed with the tests. So I didn't specifically create the osd and added out-of-band.
> 
> Why do you've to create OSDs? I guess you meant pools?
Sry, I meant pools
> 
> > @batrick
> > Do you want me to specifically add the data pool as part of the test case ?

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -302,6 +305,27 @@ def test_volume_rm(self):
+        volumes = json.loads(self._fs_cmd("volume", "ls", "--format=json-pretty"))
+        if (self.volname in [volume['name'] for volume in volumes]):
+            self.fail("Expected the 'fs volume rm' command to succeed. "
+                      "The volume {0} not removed.".format(self.volname))
```
Try to prefer using `self.assertFoo` methods. You can restructure as:

````
volnames = [volume['name'] for volume in volumes]
self.assertNotIn(self.volname, volnames)
````

This gives much better error messages if the check fails.

kotreshhr replied:
Done
